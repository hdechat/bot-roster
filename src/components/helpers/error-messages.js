export const missingName = 'Team name is required!';
export const badNumberOfStarters = 'There must be 10 starters on your team!';
export const badNumberOfSubs = 'There must be 5 subs on your team!';
export const badScore = 'Robot is too strong! Maximum allowable Total Attribute Score is 100';
export const duplicateScore = 'Robot has the same Total Attribute Score as another team member. Every team member must have a unique Total Attribute Score.';
export const duplicateFirstName = 'Robot has the same First Name as another team member. Every team member must have a unique First Name.';
export const duplicateLastName = 'Robot has the same Last Name as another team member. Every team member must have a unique Last Name.';
export const maxPlayers = "Maximum players in category!";
