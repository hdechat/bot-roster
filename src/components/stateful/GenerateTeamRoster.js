import React, { Component } from 'react';
import AddRobot from './AddRobot';
import Roster from '../stateless/Roster';
import * as errors from '../helpers/error-messages';

const initialState = {
  teamName: '',
  starters: [],
  subs: [],
  error: ''
};

class GenerateTeamRoster extends Component {
  constructor(props) {
    super(props);

    this.state = initialState;

    this.count = 0;
  };

  submitTeamRoster = (event) => {
    event.preventDefault();

    let rosterStatus = this.validateTeamRoster(this.state);

    if (rosterStatus === 'valid') {
      this.props.addTeamRoster({...this.state, error: ''});
      this.clearStateValues();
      this.count = 0;
    } else {
      this.setState({error: rosterStatus})
    }
  };

  validateTeamRoster({ teamName, starters, subs }) {
    if (!teamName) {
      return errors.missingName;
    } else if (starters.length !== 10) {
      return errors.badNumberOfStarters;
    } else if (subs.length !== 5) {
      return errors.badNumberOfSubs;
    } else {
      return 'valid';
    }
  }

  clearStateValues = () => {
    this.setState({ ...this.state, ...initialState });
  };

  deleteRobot = (robot) => {
    const { category, id } = robot;

    const filtered = this.state[category].filter(bot => bot.id !== id);

    this.setState({ [category]: filtered });
  };

  addRobotToTeam = (robot) => {
    const { category, firstName, lastName, speed, strength, agility } = robot;

    const totalAttrScore = this.calculateTotalScore(speed, strength, agility);
    const id = this.createId(firstName.slice(0, 2) + lastName.slice(0, 2));
    const newRobot = {...robot, id, totalAttrScore}

    let robotStatus = this.validateNewRobot(newRobot, totalAttrScore);

    if (robotStatus === 'valid') {
      this.setState({ [category]: [...this.state[category], newRobot], error: '' });
    } else {
      this.count--;
      this.setState({ error: robotStatus});
    }
  };

  validateNewRobot(robot, score) {
    const { firstName, lastName, category } = robot;
    let robotHasDupeStrength = this.checkForDuplicate('totalAttrScore', score);
    let robotHasDupeFirstName = this.checkForDuplicate('firstName', firstName);
    let robotHasDupeLastName = this.checkForDuplicate('lastName', lastName);
    let categoryFull = this.checkForCategoryFull(category);

    if (score > 100) {
      return errors.badScore;
    } else if (robotHasDupeStrength) {
      return errors.duplicateScore;
    } else if (robotHasDupeFirstName) {
      return errors.duplicateFirstName;
    } else if (robotHasDupeLastName) {
      return errors.duplicateLastName;
    } else if (categoryFull) {
      return errors.maxPlayers;
    } else {
      return 'valid';
    }
  };

  checkForCategoryFull(category) {
    if (category === 'starters') {
      return this.state.starters.length < 10 ? false : true;
    } else {
      return this.state.subs.length < 5 ? false : true;
    }
  }

  checkForDuplicate(prop, value) {
    const categories = ['starters', 'subs'];
    let duplicate;
    let count = 0;

    while(!duplicate && count < categories.length) {
      duplicate = this.state[categories[count]].find(bot => bot[prop] === value);
      count++;
    }

    return duplicate;
  }

  createId(prefix) {
    this.count++;
    let teamNumber = this.count.toString();

    prefix = teamNumber.length < 2 ? prefix += '0' : prefix;
    return prefix + teamNumber;
  }

  calculateTotalScore(speed, strength, agility) {
    return speed + strength + agility;
  }

  updateName = (robot) => {
    const updatedBots = this.state[robot.category].map(bot => {
      return bot.id === robot.id ? robot : bot;
    });

    this.setState({ [robot.category]: updatedBots })
  }

  render() {
    return (
      <div className="GenerateTeamRoster">
      <input
        required
        id="team-name"
        value={this.state.teamName} 
        onChange={(event) => this.setState({ teamName: event.target.value })}
        placeholder="Team Name" />
        <button className="generate__submit-roster" onClick={this.submitTeamRoster}>Submit Team Roster</button>
        <p className="generate__error-message">{this.state.error}</p>
        <div className="generate__container">
          <AddRobot addRobotToTeam={this.addRobotToTeam}/>
          <Roster deleteRobot={this.deleteRobot} updateName={this.updateName} team={this.state}/>
        </div>
      </div>
    );
  };
};

export default GenerateTeamRoster;
