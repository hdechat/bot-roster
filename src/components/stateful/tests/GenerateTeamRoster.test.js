import React from 'react';
import GenerateTeamRoster from '../GenerateTeamRoster';
import * as errors from '../../helpers/error-messages';
import { shallow, mount  } from 'enzyme';

describe('GenerateTeamRoster', () => {
  let wrapper;
  let mockAddTeamRoster;
  let mockState;

  beforeEach(() => {
    mockState = {
      teamName: '',
      starters: [],
      subs: [],
      error: ''
    }

    mockAddTeamRoster = jest.fn();

    wrapper = shallow(<GenerateTeamRoster addTeamRoster={mockAddTeamRoster} />);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('SUBMIT_TEAM_ROSTER', () => {
    const mockEvent = {preventDefault: jest.fn()};

    beforeEach(() => {
      let starters = [];
      let subs = [];

      for (let i = 0; i < 10; i++) {
        starters.push({
          id: `ABC12${i}`,
          firstName: `${i}Twikki`,
          lastName: `${i}Rogers`,
          speed: 2 + i,
          strength: 3 + i,
          agility: 4 + i,
          totalAttrScore: 9 + i
        });
      }

      for (let i = 0; i < 5; i++) {
        subs.push({
          id: `DEF12${i}`,
          firstName: `${i}Kki`,
          lastName: `${i}Gers`,
          speed: 12 + i,
          strength: 13 + i,
          agility: 14 + i,
          totalAttrScore: 19 + i
        });
      }

      mockState = {
        teamName: 'BuzzKillers',
        starters,
        subs,
        error: ''
      }
    });

    it('calls validateTeamRoster', () => {
      wrapper = mount(<GenerateTeamRoster addTeamRoster={mockAddTeamRoster} />);
      wrapper.setState(mockState);
      const spy = spyOn(wrapper.instance(), 'validateTeamRoster');

      wrapper.instance().forceUpdate();
      wrapper.instance().submitTeamRoster(mockEvent);

      expect(spy).toHaveBeenCalled();
    });

    it('calls addTeamRoster with the correct params', () => {
      wrapper.setState(mockState);
      wrapper.instance().submitTeamRoster(mockEvent);

      expect(mockAddTeamRoster).toHaveBeenCalledWith(mockState);
    });

    it('it calls clearStateValues', () => {
      wrapper = mount(<GenerateTeamRoster addTeamRoster={mockAddTeamRoster} />);
      wrapper.setState(mockState);
      const spy = spyOn(wrapper.instance(), 'clearStateValues');

      wrapper.instance().forceUpdate();
      wrapper.instance().submitTeamRoster(mockEvent);

      expect(spy).toHaveBeenCalled();
    });

    it('resets count to 0', () => {
      wrapper.setState(mockState);
      wrapper.instance().count = 1

      wrapper.instance().submitTeamRoster(mockEvent);

      expect(wrapper.instance().count).toEqual(0);
    });

    it('changes state.error and renders error message if user submits roster without a name', () => {
      let starters = [];
      let subs = [];

      for (let i = 0; i < 10; i++) {
        starters.push({robot: i})
      }

      for (let i = 0; i < 5; i++) {
        subs.push({robot: i})
      }

      mockState = {
        teamName: '',
        starters,
        subs,
        error: ''
      }

      wrapper.setState(mockState);
      wrapper.instance().submitTeamRoster(mockEvent);

      expect(wrapper.state('error')).toEqual(errors.missingName);
      expect(wrapper.find('.error-message').text()).toEqual(errors.missingName);

    });

    it('changes state.error and renders error message if user submits roster that does not have 10 starters', () => {
      let starters = [];
      let subs = [];

      for (let i = 0; i < 8; i++) {
        starters.push({robot: i})
      }

      for (let i = 0; i < 5; i++) {
        subs.push({robot: i})
      }

      mockState = {
        teamName: 'BuzzKillers',
        starters,
        subs,
        error: ''
      }

      wrapper.setState(mockState);
      wrapper.instance().submitTeamRoster(mockEvent);

      expect(wrapper.state('error')).toEqual(errors.badNumberOfStarters);
      expect(wrapper.find('.error-message').text()).toEqual(errors.badNumberOfStarters)
    });

    it('changes state.error and renders error message if user submits roster that does not have 5 subs', () => {
      let starters = [];
      let subs = [];

      for (let i = 0; i < 10; i++) {
        starters.push({robot: i})
      }

      for (let i = 0; i < 3; i++) {
        subs.push({robot: i})
      }

      mockState = {
        teamName: 'BuzzKillers',
        starters,
        subs,
        error: ''
      }

      wrapper.setState(mockState);
      wrapper.instance().submitTeamRoster(mockEvent);

      expect(wrapper.state('error')).toEqual(errors.badNumberOfSubs);
      expect(wrapper.find('.error-message').text()).toEqual(errors.badNumberOfSubs)
    });
  });

  describe('VALIDATE_TEAM_ROSTER', () => {
    beforeEach(() => {
      let starters = [];
      let subs = [];

      for (let i = 0; i < 10; i++) {
        starters.push({
          id: `ABC12${i}`,
          firstName: `${i}Twikki`,
          lastName: `${i}Rogers`,
          speed: 2 + i,
          strength: 3 + i,
          agility: 4 + i,
          totalAttrScore: 9 + i
        });
      }

      for (let i = 0; i < 5; i++) {
        subs.push({
          id: `DEF12${i}`,
          firstName: `${i}Kki`,
          lastName: `${i}Gers`,
          speed: 12 + i,
          strength: 13 + i,
          agility: 14 + i,
          totalAttrScore: 19 + i
        });
      }

      mockState = {
        teamName: 'BuzzKillers',
        starters,
        subs,
        error: ''
      }
    });

    it('should return missing name error string if teamName is undefined', () => {
      const mockStateMissingName = {...mockState, teamName: ''};

      const actual = wrapper.instance().validateTeamRoster(mockStateMissingName);
      const expected = errors.missingName;

      expect(actual).toEqual(expected);
    });

    it('should return bad number of starters error string if there are not 10 starters on the team', () => {
      const mockStateBadNumberOfStarters = {...mockState, starters: []};

      const actual = wrapper.instance().validateTeamRoster(mockStateBadNumberOfStarters);
      const expected = errors.badNumberOfStarters;

      expect(actual).toEqual(expected);
    });

    it('should return bad number of subs error string if there are not 5 starters on the team', () => {
      const mockStateBadNumberOfSubs = {...mockState, subs: []};

      const actual = wrapper.instance().validateTeamRoster(mockStateBadNumberOfSubs);
      const expected = errors.badNumberOfSubs;

      expect(actual).toEqual(expected);
    });

    it('should return valid if all parameters pass the conditions', () => {
      expect(wrapper.instance().validateTeamRoster(mockState)).toEqual('valid');
    });
  });

  describe('CLEAR_STATE_VALUES', () => {
    it('should clear state values', () => {
      wrapper.setState({ teamName: 'Buzz Killers'});

      wrapper.instance().clearStateValues();

      expect(wrapper.state()).toEqual(mockState);
    });
  });

  describe('ADD_ROBOT_TO_TEAM', () => {
    const mockRobotSub = {
      category: 'subs',
      firstName: 'r5',
      lastName: 'd4',
      speed: 10,
      strength: 40,
      agility: 30  
    }

    const mockRobotStarter = {
      category: 'starters',
      firstName: 'Rosie',
      lastName: 'Robot',
      speed: 10,
      strength: 40,
      agility: 50 
    }

    const mockRobotTooStrong = {
      category: 'starters',
      firstName: 'Voltron',
      lastName: 'Jackson',
      speed: 30,
      strength: 40,
      agility: 50 
    }

    it('calls calculateTotalScore with the correct params', () => {
      wrapper = mount(<GenerateTeamRoster addTeamRoster={mockAddTeamRoster} />);
      const spy = spyOn(wrapper.instance(), 'calculateTotalScore');

      wrapper.instance().forceUpdate();
      wrapper.instance().addRobotToTeam(mockRobotSub);

      expect(spy).toHaveBeenCalledWith(10, 40, 30);
    });

    it('calls createId with the correct params', () => {
      wrapper = mount(<GenerateTeamRoster addTeamRoster={mockAddTeamRoster} />);
      const spy = spyOn(wrapper.instance(), 'createId');

      wrapper.instance().forceUpdate();
      wrapper.instance().addRobotToTeam(mockRobotSub);

      expect(spy).toHaveBeenCalledWith('r5d4');
    });

    it('calls validateNewRobot with the correct params', () => {
      wrapper = mount(<GenerateTeamRoster addTeamRoster={mockAddTeamRoster} />);
      const spy = spyOn(wrapper.instance(), 'validateNewRobot');

      wrapper.instance().forceUpdate();
      wrapper.instance().addRobotToTeam(mockRobotSub);

      const mockRobot = {...mockRobotSub, id: 'r5d401', totalAttrScore: 80}
      expect(spy).toHaveBeenCalledWith(mockRobot, 80)
    });

    it('resets state to the correct props with the updated robot data', () => {
      const mockUpdatedState = {
        teamName: '',
        starters: [{...mockRobotStarter, id: 'RoRo02', totalAttrScore: 100}],
        subs: [{...mockRobotSub, id: 'r5d401', totalAttrScore: 80}],
        error: ''
      }

      wrapper.instance().addRobotToTeam(mockRobotSub);
      wrapper.instance().addRobotToTeam(mockRobotStarter);

      expect(wrapper.state()).toEqual(mockUpdatedState);
    });

    it('does not change count if robotStatus is not valid', () => {
        wrapper.count = 1;

        wrapper.instance().addRobotToTeam(mockRobotTooStrong);

        expect(wrapper.count).toEqual(1)
    });

    it('changes state.error and renders error message if totalAttrScore is greater than 100', () => {
      wrapper.instance().addRobotToTeam(mockRobotTooStrong);

      expect (wrapper.state()).toEqual({...mockState, error: errors.badScore})
      expect(wrapper.find('.error-message').text()).toEqual(errors.badScore)
    });

    it('changes state.error and renders error message if totalAttrScore is a duplicate', () => {
      const mockUpdatedState = {
        teamName: '',
        starters: [{...mockRobotStarter, id: 'RoRo02', totalAttrScore: 100}],
        subs: [{...mockRobotSub, id: 'r5d401', totalAttrScore: 80}],
        error: ''
      };

      const mockRobotDupe = {
        category: 'starters',
        firstName: 'Twiki',
        lastName: 'Rogers',
        speed: 20,
        strength: 30,
        agility: 30,
        error: ''
      };

      wrapper.setState(mockUpdatedState);
      wrapper.instance().addRobotToTeam(mockRobotDupe);

      expect(wrapper.state()).toEqual({...mockUpdatedState, error: errors.duplicateScore});
      expect(wrapper.find('.error-message').text()).toEqual(errors.duplicateScore);
    });

    it('changes state.error and renders error message if firstName is a duplicate in the team', () => {
      const mockUpdatedState = {
        teamName: '',
        starters: [{...mockRobotStarter, id: 'RoRo02', totalAttrScore: 100}],
        subs: [{...mockRobotSub, id: 'r5d401', totalAttrScore: 80}],
        error: ''
      };

      const mockRobotDupe = {
        category: 'starters',
        firstName: 'r5',
        lastName: 'Rogers',
        speed: 2,
        strength: 3,
        agility: 3,
        error: ''
      };

      wrapper.setState(mockUpdatedState);
      wrapper.instance().addRobotToTeam(mockRobotDupe);

      expect(wrapper.state()).toEqual({...mockUpdatedState, error: errors.duplicateFirstName});
      expect(wrapper.find('.error-message').text()).toEqual(errors.duplicateFirstName);
    });

    it('changes state.error and renders error message if lastName is a duplicate in the team', () => {
      const mockUpdatedState = {
        teamName: '',
        starters: [{...mockRobotStarter, id: 'RoRo02', totalAttrScore: 100}],
        subs: [{...mockRobotSub, id: 'r5d401', totalAttrScore: 80}],
        error: ''
      };

      const mockRobotDupe = {
        category: 'starters',
        firstName: 'Twiki',
        lastName: 'd4',
        speed: 2,
        strength: 3,
        agility: 3,
        error: ''
      };

      wrapper.setState(mockUpdatedState);
      wrapper.instance().addRobotToTeam(mockRobotDupe);

      expect(wrapper.state()).toEqual({...mockUpdatedState, error: errors.duplicateLastName});
      expect(wrapper.find('.error-message').text()).toEqual(errors.duplicateLastName);
    });
  });

  describe('VALIDATE_NEW_ROBOT', () => {
    let mockRobot;
    let mockRobotSub;
    let mockRobotStarter;
    let mockUpdatedState;
    let mockState;

    beforeEach(() => {
      mockRobotSub = {
        category: 'subs',
        firstName: 'r5',
        lastName: 'd4',
        speed: 10,
        strength: 40,
        agility: 30  
      }

      mockRobotStarter = {
        category: 'starters',
        firstName: 'Rosie',
        lastName: 'Robot',
        speed: 10,
        strength: 40,
        agility: 50 
      }

      mockUpdatedState = {
        teamName: '',
        starters: [{...mockRobotStarter, id: 'RoRo02', totalAttrScore: 100}],
        subs: [{...mockRobotSub, id: 'r5d401', totalAttrScore: 80}],
        error: ''
      };

      mockRobot = {
        category: 'subs',
        firstName: 'Twikki',
        lastName: 'Bell',
        speed: 10,
        strength: 50,
        agility: 30  
      }

      let starters = [];
      let subs = [];

      for (let i = 0; i < 10; i++) {
        starters.push({
          id: `ABC12${i}`,
          firstName: `${i}Twikki`,
          lastName: `${i}Rogers`,
          speed: 2 + i,
          strength: 3 + i,
          agility: 4 + i,
          totalAttrScore: 9 + i
        });
      }

      for (let i = 0; i < 5; i++) {
        subs.push({
          id: `DEF12${i}`,
          firstName: `${i}Kki`,
          lastName: `${i}Gers`,
          speed: 12 + i,
          strength: 13 + i,
          agility: 14 + i,
          totalAttrScore: 19 + i
        });
      }

      mockState = {
        teamName: 'BuzzKillers',
        starters,
        subs,
        error: ''
      };
    });

    it('calls checkForDuplicate 3 times', () => {
      wrapper = mount(<GenerateTeamRoster addTeamRoster={mockAddTeamRoster} />);
      const spy = spyOn(wrapper.instance(), 'checkForDuplicate');

      wrapper.instance().forceUpdate()
      wrapper.instance().validateNewRobot(mockRobot, 80)

      expect(spy).toHaveBeenCalledTimes(3)
    });

    it('returns correct error string if score is > 100', () => {
      const actual = wrapper.instance().validateNewRobot(mockRobot, 200);
      const expected = errors.badScore;

      expect(actual).toEqual(expected);
    });

    it('returns correct error string if score is not unique', () => {
      wrapper.setState(mockUpdatedState);

      const actual = wrapper.instance().validateNewRobot(mockRobot, 80);
      const expected = errors.duplicateScore;

      expect(actual).toEqual(expected);
    });

    it('returns correct error string if firstName is not unique', () => {
      wrapper.setState(mockUpdatedState);

      const actual = wrapper.instance().validateNewRobot({...mockRobot, firstName: 'Rosie'}, 90);
      const expected = errors.duplicateFirstName;

      expect(actual).toEqual(expected);
    });

    it('returns correct error string if lastName is not unique', () => {
      wrapper.setState(mockUpdatedState);

      const actual = wrapper.instance().validateNewRobot({...mockRobot, lastName: 'Robot'}, 90);
      const expected = errors.duplicateLastName;
      
      expect(actual).toEqual(expected);
    });

    it('returns correct errors string if there are maximum number of starters', () => {
      wrapper.setState(mockState);

      const actual = wrapper.instance().validateNewRobot({...mockRobot, category: 'starters'}, 90);
      const expected = errors.maxPlayers;

      expect(actual).toEqual(expected);
    });

    it('returns correct errors string if there are maximum number of subs', () => {
      wrapper.setState(mockState);

      const actual = wrapper.instance().validateNewRobot(mockRobot, 90);
      const expected = errors.maxPlayers;

      expect(actual).toEqual(expected);
    });

    it('returns valid if none of the checkForDuplicate returns return a result', () => {
      const actual = wrapper.instance().validateNewRobot(mockRobot, 80);

      expect(actual).toEqual('valid');
    });
  });

  describe('CHECK_FOR_CATEGORY_FULL', () => {
    beforeEach(() => {
      let starters = [];
      let subs = [];

      for (let i = 0; i < 10; i++) {
        starters.push({
          id: `ABC12${i}`,
          firstName: `${i}Twikki`,
          lastName: `${i}Rogers`,
          speed: 2 + i,
          strength: 3 + i,
          agility: 4 + i,
          totalAttrScore: 9 + i
        });
      }

      for (let i = 0; i < 5; i++) {
        subs.push({
          id: `DEF12${i}`,
          firstName: `${i}Kki`,
          lastName: `${i}Gers`,
          speed: 12 + i,
          strength: 13 + i,
          agility: 14 + i,
          totalAttrScore: 19 + i
        });
      }

      mockState = {
        teamName: 'BuzzKillers',
        starters,
        subs,
        error: ''
      };
    });

    it('should return true if there are 10 starters', () => {
      wrapper.setState(mockState);

      const actual = wrapper.instance().checkForCategoryFull('starters');

      expect(actual).toEqual(true);
    });

    it('should return true if there are 5 subs', () => {
      wrapper.setState(mockState);

      const actual = wrapper.instance().checkForCategoryFull('subs');

      expect(actual).toEqual(true);
    });

    it('should return false if there are less than 10 starters', () => {
      wrapper.setState({...mockState, starters: []});

      const actual = wrapper.instance().checkForCategoryFull('starters');

      expect(actual).toEqual(false);
    });

    it('should return false if there are less than 5 subs', () => {
      wrapper.setState({...mockState, subs: []});

      const actual = wrapper.instance().checkForCategoryFull('subs')

      expect(actual).toEqual(false)
    });
  });

  describe('CHECK_FOR_DUPLICATE', () => {
    let mockRobot;

    beforeEach(() => {
      mockRobot = {
        category: 'subs',
        firstName: 'Twikki',
        lastName: 'Bell',
        speed: 10,
        strength: 50,
        agility: 30  
      }

      wrapper.setState({...mockState, starters: [mockRobot]});
    });

    it('returns undefined if there is not duplicate found', () => {
      const actual = wrapper.instance().checkForDuplicate('firstName', 'Rosie');

      expect(actual).toEqual(undefined);
    });

    it('returns the duplicate value if found', () => {
      const actual = wrapper.instance().checkForDuplicate('firstName', 'Twikki');

      expect(actual).toEqual(mockRobot);
    });
  });

  describe('CREATE_ID', () => {
    it('returns prefix and teamNumber with a zero if the count is a single digit', () => {
    const mockPrefix = 'AbCd';
    const actual = wrapper.instance().createId(mockPrefix);
    const expectation = 'AbCd01';

    expect(actual).toEqual(expectation);
    }); 

    it('returns prefix and teamNumber and does not add a zero if the count is double digit', () => {
      const mockPrefix = 'AbCd';
      for (let i = 0; i < 10; i++) {
        wrapper.instance().createId(mockPrefix);
      }
      const actual = wrapper.instance().createId(mockPrefix);
      const expectation = 'AbCd11';

      expect(actual).toEqual(expectation);
    });
  });

  describe('CALCULATE_TOTAL_SCORE', () => {
    it('returns the sum of the args', () => {
      const actual = wrapper.instance().calculateTotalScore(1, 2, 3);

      expect(actual).toEqual(6);
    }); 
  });

  describe('UPDATE_NAME_IN_GENERATE_ROSTER', () => {
    let mockRobot = {
      category: 'subs',
      firstName: 'Twikki',
      lastName: 'Bell',
      speed: 10,
      strength: 50,
      agility: 30  
    }

    let mockState = {
      teamName: 'Buzzers',
      starters: [],
      subs: [mockRobot],
      error: ''
    };

    it('updates robot in state', () => {
      wrapper.setState(mockState);
      wrapper.instance().updateName({...mockRobot, firstName: 'Tinker'});

      const expected = {
        teamName: 'Buzzers',
        starters: [],
        subs: [{...mockRobot, firstName: 'Tinker'}],
        error: ''
      };
      expect(wrapper.state()).toEqual(expected)
    });
  });

  it('renders all the expected elements', () => {
    expect(wrapper.find('div.GenerateTeamRoster').length).toEqual(1);
    expect(wrapper.find('input').length).toEqual(1);
    expect(wrapper.find('button').length).toEqual(1);
    expect(wrapper.find('p.error-message').length).toEqual(1);
    expect(wrapper.find('AddRobot').length).toEqual(1);
    expect(wrapper.find('Roster').length).toEqual(1);
  });

  it('resets teamName state with input onChange', () => {
    expect(wrapper.state('teamName')).toBe('');

    const mockEvent = {target: {value: 'Tw'}};

    wrapper.find('input').simulate('change', mockEvent);

    expect(wrapper.state('teamName')).toBe('Tw')
  });

  it('calls submitTeamRoster with button click', () => {
    wrapper = mount(<GenerateTeamRoster addTeamRoster={mockAddTeamRoster} />);
    const spy = spyOn(wrapper.instance(), 'submitTeamRoster');

    wrapper.instance().forceUpdate();
    wrapper.find('button.submit-roster').simulate('click');

    expect(spy).toHaveBeenCalled();
  });
});
