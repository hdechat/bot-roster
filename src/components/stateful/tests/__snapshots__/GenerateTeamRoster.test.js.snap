// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateTeamRoster matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GenerateTeamRoster
    addTeamRoster={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          id="team-name"
          onChange={[Function]}
          placeholder="Team Name"
          required={true}
          value=""
/>,
        <button
          className="generate__submit-roster"
          onClick={[Function]}
>
          Submit Team Roster
</button>,
        <p
          className="generate__error-message"
/>,
        <div
          className="generate__container"
>
          <AddRobot
                    addRobotToTeam={[Function]}
          />
          <Roster
                    team={
                              Object {
                                        "error": "",
                                        "starters": Array [],
                                        "subs": Array [],
                                        "teamName": "",
                                      }
                    }
                    updateName={[Function]}
          />
</div>,
      ],
      "className": "GenerateTeamRoster",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "id": "team-name",
          "onChange": [Function],
          "placeholder": "Team Name",
          "required": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Submit Team Roster",
          "className": "generate__submit-roster",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Submit Team Roster",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "generate__error-message",
        },
        "ref": null,
        "rendered": "",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <AddRobot
              addRobotToTeam={[Function]}
/>,
            <Roster
              team={
                            Object {
                                          "error": "",
                                          "starters": Array [],
                                          "subs": Array [],
                                          "teamName": "",
                                        }
              }
              updateName={[Function]}
/>,
          ],
          "className": "generate__container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "addRobotToTeam": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "team": Object {
                "error": "",
                "starters": Array [],
                "subs": Array [],
                "teamName": "",
              },
              "updateName": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            id="team-name"
            onChange={[Function]}
            placeholder="Team Name"
            required={true}
            value=""
/>,
          <button
            className="generate__submit-roster"
            onClick={[Function]}
>
            Submit Team Roster
</button>,
          <p
            className="generate__error-message"
/>,
          <div
            className="generate__container"
>
            <AddRobot
                        addRobotToTeam={[Function]}
            />
            <Roster
                        team={
                                    Object {
                                                "error": "",
                                                "starters": Array [],
                                                "subs": Array [],
                                                "teamName": "",
                                              }
                        }
                        updateName={[Function]}
            />
</div>,
        ],
        "className": "GenerateTeamRoster",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "team-name",
            "onChange": [Function],
            "placeholder": "Team Name",
            "required": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit Team Roster",
            "className": "generate__submit-roster",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Submit Team Roster",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "generate__error-message",
          },
          "ref": null,
          "rendered": "",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <AddRobot
                addRobotToTeam={[Function]}
/>,
              <Roster
                team={
                                Object {
                                                "error": "",
                                                "starters": Array [],
                                                "subs": Array [],
                                                "teamName": "",
                                              }
                }
                updateName={[Function]}
/>,
            ],
            "className": "generate__container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "addRobotToTeam": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "team": Object {
                  "error": "",
                  "starters": Array [],
                  "subs": Array [],
                  "teamName": "",
                },
                "updateName": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
